### ISSUE TITLE
Can't add more then 8 network interfaces / ips 

### ISSUE THREAD

**Issue by koboltmarky** [bpg/terraform-provider-proxmox#798]
**Error**
When we try to add a ninth interface we get the following error:

```
│ Error: Too many network_device blocks

│ 

│   on ../../modules/proxmox/resource_vm.tf line 34, in resource "proxmox_virtual_environment_vm" "xxxxx":

│   34:     content {
│ 
│ No more than 8 "network_device" blocks are allowed
╵
╷
│ Error: Too many ip_config blocks
│ 
│   on ../../modules/proxmox/resource_vm.tf line 48, in resource "proxmox_virtual_environment_vm" "xxxx":
│   48:       content {
│ 
│ No more than 8 "ip_config" blocks are allowed
```

I can add an interface through the ui.
There was a limitation to 8 interface in proxmox in the past. Afaik it is now 32.

https://forum.proxmox.com/threads/proxmox-vm-maxium-number-of-network-interfaces.64788/

**To Reproduce**
Add a vm with 9 interfaces.

Our code:

```hcl
resource "proxmox_virtual_environment_vm" "XXXX" {
  name      = var.name
  node_name = var.proxmox_node
  migrate   = true // migrate the VM on node change
  tags      = var.tags

  clone {
    vm_id = var.vm_id
  }

  memory {
    dedicated = var.memory
  }

  # Not working
  #boot_order = [2, 1]

  operating_system {
    type = "l26"
  }

  agent {
    enabled = true
  }

  cpu {
    sockets = 2
    cores   = 1
    flags   = []
  }

  dynamic "network_device" {
    for_each = var.network_devices
    content {
      bridge  = network_device.value["bridge"]
      vlan_id = network_device.value["vlan_id"]
      mtu     = network_device.value["mtu"]
    }
  }

  initialization {
    dns {
      server = var.dns_server
    }

    dynamic "ip_config" {
      for_each = var.ip_configs
      content {
        ipv4 {
          address = ip_config.value["address"]
          gateway = ip_config.value["gateway"]
        }
      }
    }

    datastore_id = var.datastore_id

    user_account {
      keys     = var.ssh_keys
      username = var.username
    }
  }

  disk {
    size         = var.disk_size
    datastore_id = var.datastore_id
    interface    = "scsi0"
  }

  timeout_clone    = 120
  timeout_start_vm = 120
  timeout_create   = 120

  lifecycle {
    ignore_changes = [
      initialization[0].user_account[0].keys, # We need this to avoid a recreation of the vm when the ssh keys are changed
    ]
  }
}
```

Working example with 8 interfaces.
We use the vm code in module. We call the module with the following code:
 
```
module "XXXX" {
  source       = "../../modules/proxmox"
  proxmox_node = "xxxxx"
  tags         = ["ubuntu22.04", "network", "xxxxx"]
  cores        = 2
  sockets      = 2
  memory       = 2048
  name         = "xxxx"
  dns_server   = "192.168.xx.xx"
  ip_configs = [
    {
      address = "192.168.x.x",
      gateway = "192.168.x.x"
    },
    {
      address = "10.6.x.x/24",
      gateway = ""
    },
    {.
      address = "10.6.x.x/28"
      gateway = ""
    },
    {
      address = "10.6.x.x/28"
      gateway = ""
    },
    {
      address = "10.6.x.x/28"
      gateway = ""
    },
    {
      address = "10.6.x.x/28"
      gateway = ""
    },
    {
      address = "10.6.x.x/28"
      gateway = ""
    },
    {
      address = "10.6.x.x/28"
      gateway = ""
    }
  ]
  datastore_id = "local"
  username     = "ubuntu"
  ssh_keys     = local.ssh_keys
  network_devices = [
    {
      bridge  = "vmbr1"
      vlan_id = "4001"
      mtu     = 1500
    },
    {
      bridge  = "vmbr1"
      vlan_id = "101"
      mtu     = 1500
    },
    {
      bridge  = "vmbr1"
      vlan_id = "1001"
      mtu     = 1500
    },
    {
      bridge  = "vmbr1"
      vlan_id = "1004"
      mtu     = 1500
    },
    {
      bridge  = "vmbr1"
      vlan_id = "1005"
      mtu     = 1500
    },
    {
      bridge  = "vmbr1"
      vlan_id = "1011"
      mtu     = 1500
    },
    {
      bridge  = "vmbr1"
      vlan_id = "1012"
      mtu     = 1500
    },
    {
      bridge  = "vmbr1"
      vlan_id = "1013"
      mtu     = 1500
    }
  ]
  disk_size = 20
  vm_id     = "9001"
}
```

**Expected behavior**
I can add more the 8 interfaces and ips. The limitation should be the proxmox limitation.

**Versions**
- Single or clustered Proxmox: 3x Nodes Cluster
- Provider version (ideally it should be the latestversion): 
- Terraform version:
- OS (where you run Terraform from): Ubunutu 20.04

**Comment by bpg** [comment]
Hi @koboltmarky! 👋🏼

Indeed, there is a [hardcoded limit](https://github.com/bpg/terraform-provider-proxmox/blob/v0.40.0/proxmox/nodes/vms/vms_types.go#L473-L480) that needs to be extended. However, instead of copy-pasting struct fields up to 32 🤦🏼‍♂️, we should consider implementing a dynamic mapping.

I would appreciate some help on this ticket, as, frankly speaking, this feature is not currently at the top of my TODO list.

**Comment by bitchecker** [comment]
Do you think that the “stupid version” can help to unlock a hard-coded limit that should be updated?

Maybe the “code refactor” can be done in the future.

I'm not a developer, but I tried to take a look at the code anyway, the updates should be done at least: 
- [here](https://github.com/bpg/terraform-provider-proxmox/blob/main/proxmox/nodes/vms/vms_types.go) 
- [here](https://github.com/bpg/terraform-provider-proxmox/blob/main/proxmoxtf/resource/vm.go)

I don't know the project, but if you think that can be helpful, I can try to run a Pull Request.

**Comment by bitchecker** [comment]
Ciao @bpg,
I'm not a developer, but I really like this project and I would like to try to help.

About this issue, after your suggestion that I used as entry point for my analysis, I forked the project and updated the code in order to remove the limitation.

This is the “stupid” version, no dynamic mapping (is the first time that I work on go), but maybe in the meantime this simple [PR](https://github.com/bpg/terraform-provider-proxmox/pull/822) can help to unlock a problem.
